File: exercises/cs-core/traceable-priority-shortest-path-lab/cpp/tests/test_lab.cpp
#include "traceable_priority_shortest_path_lab/lab.hpp"

#include <cassert>
#include <cstdint>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

using namespace traceable_priority_shortest_path_lab;

int main() {
  TraceableMinHeap heap;
  std::size_t comparisons = 0;
  std::size_t swaps = 0;
  for (const auto value : {7, 3, 9, 1, 5}) {
    const auto event = heap.push(value);
    comparisons += event.comparisons;
    swaps += event.swaps;
  }
  assert((heap.values() == std::vector<std::int64_t>{1, 3, 9, 7, 5}));
  assert(comparisons == 5 && swaps == 3);
  const auto popped = heap.pop_min();
  assert(popped.value == 1 && popped.comparisons == 3 && popped.swaps == 1);
  assert(is_min_heap(build_min_heap({9, -1, 7, -1, 3, 2}).values));
  assert(!is_min_heap({2, 1, 3}));

  StableMinPriorityQueue empty_queue;
  try {
    static_cast<void>(empty_queue.pop());
    assert(false);
  } catch (const std::out_of_range&) {
    assert(empty_queue.empty());
  }
  const auto drained = drain_by_priority({{"same", 1}, {"same", 1}, {"urgent", -1}, {"later", 2}});
  assert(drained[0].label == "urgent");
  assert(drained[1].sequence == 0 && drained[2].sequence == 1);

  const auto graph = sample_weighted_graph();
  const auto trace = dijkstra(graph, 0);
  assert((trace.settled == std::vector<std::size_t>{0, 2, 1, 3, 4, 5}));
  assert(trace.edge_checks == 16 && trace.relaxations == 8);
  assert(trace.queue_pushes == 9 && trace.stale_pops == 3 && trace.max_frontier == 4);
  assert((shortest_path(graph, 0, 5).path == std::vector<std::size_t>{0, 2, 1, 3, 4, 5}));
  assert(shortest_path(graph, 0, 6).path.empty());
  assert((vertices_within_distance(graph, 0, 4) == std::vector<std::size_t>{0, 2, 1, 3}));

  try {
    static_cast<void>(WeightedGraph(2, {{0, 1, -1}}));
    assert(false);
  } catch (const std::invalid_argument&) {
  }
  try {
    const auto overflow_graph = WeightedGraph(
        3, {{0, 1, std::numeric_limits<std::int64_t>::max()}, {1, 2, 1}});
    static_cast<void>(dijkstra(overflow_graph, 0));
    assert(false);
  } catch (const std::overflow_error&) {
  }

  assert(build_heap_report().find("comparisons=5，swaps=3") != std::string::npos);
  assert(build_queue_report().find("equal_priority_fifo=yes") != std::string::npos);
  assert(build_dijkstra_report().find("stale_pops=3，max_frontier=4") != std::string::npos);
  return 0;
}

File: exercises/cs-core/traceable-priority-shortest-path-lab/python/tests/test_lab.py
from __future__ import annotations

import unittest

from traceable_priority_shortest_path_lab import (
    StableMinPriorityQueue, TraceableMinHeap, WeightedGraph, build_dijkstra_report,
    build_heap_report, build_min_heap, build_queue_report, dijkstra, drain_by_priority,
    is_min_heap, sample_weighted_graph, shortest_path, vertices_within_distance,
)
from traceable_priority_shortest_path_lab.weighted import MAX_DISTANCE


class HeapTests(unittest.TestCase):
    def test_fixed_trace_and_underflow(self) -> None:
        heap = TraceableMinHeap()
        events = [heap.push(value) for value in (7, 3, 9, 1, 5)]
        self.assertEqual(heap.values, (1, 3, 9, 7, 5))
        self.assertEqual(sum(event.comparisons for event in events), 5)
        self.assertEqual(sum(event.swaps for event in events), 3)
        popped = heap.pop_min()
        self.assertEqual((popped.value, popped.values, popped.comparisons, popped.swaps), (1, (3, 5, 9, 7), 3, 1))
        while not heap.empty:
            heap.pop_min()
        with self.assertRaises(IndexError):
            heap.peek_min()
        self.assertEqual(heap.values, ())

    def test_build_and_invariant(self) -> None:
        original = [9, -1, 7, -1, 3, 2]
        trace = build_min_heap(original)
        self.assertTrue(is_min_heap(trace.values))
        self.assertFalse(is_min_heap((2, 1, 3)))
        self.assertEqual(original, [9, -1, 7, -1, 3, 2])


class QueueTests(unittest.TestCase):
    def test_stable_order_and_input_unchanged(self) -> None:
        tasks = [("same", 1), ("same", 1), ("urgent", -1), ("later", 2)]
        result = drain_by_priority(tasks)
        self.assertEqual([(item.label, item.priority, item.sequence) for item in result], [
            ("urgent", -1, 2), ("same", 1, 0), ("same", 1, 1), ("later", 2, 3),
        ])
        self.assertEqual(tasks[0], ("same", 1))

    def test_underflow_is_unchanged(self) -> None:
        queue = StableMinPriorityQueue()
        with self.assertRaises(IndexError):
            queue.pop()
        self.assertTrue(queue.empty)


class DijkstraTests(unittest.TestCase):
    def test_fixed_trace_and_path(self) -> None:
        graph = sample_weighted_graph()
        trace = dijkstra(graph, 0)
        self.assertEqual(trace.settled, (0, 2, 1, 3, 4, 5))
        self.assertEqual(trace.distances, (0, 3, 1, 4, 7, 8, None))
        self.assertEqual(trace.parents, (None, 2, 0, 1, 3, 4, None))
        self.assertEqual((trace.edge_checks, trace.relaxations, trace.queue_pushes, trace.stale_pops, trace.max_frontier), (16, 8, 9, 3, 4))
        self.assertEqual(shortest_path(graph, 0, 5).path, (0, 2, 1, 3, 4, 5))
        self.assertEqual(shortest_path(graph, 0, 6).path, ())
        self.assertEqual(vertices_within_distance(graph, 0, 4), (0, 2, 1, 3))

    def test_validation_and_overflow(self) -> None:
        invalid = [((0, 0, 1),), ((0, 2, 1),), ((0, 1, -1),)]
        for edges in invalid:
            with self.assertRaises(ValueError):
                WeightedGraph(2, edges)
        with self.assertRaises(ValueError):
            WeightedGraph(2, ((0, 1, 1), (1, 0, 2)))
        graph = WeightedGraph(3, ((0, 1, MAX_DISTANCE), (1, 2, 1)))
        with self.assertRaises(OverflowError):
            dijkstra(graph, 0)

    def test_zero_weight_and_equal_distance(self) -> None:
        graph = WeightedGraph(4, ((0, 1, 0), (0, 2, 1), (1, 2, 1)))
        trace = dijkstra(graph, 0)
        self.assertEqual(trace.distances, (0, 0, 1, None))
        self.assertEqual(trace.parents[2], 0)


class ReportingTests(unittest.TestCase):
    def test_reports(self) -> None:
        self.assertIn("comparisons=5，swaps=3", build_heap_report())
        self.assertIn("equal_priority_fifo=yes", build_queue_report())
        self.assertIn("stale_pops=3，max_frontier=4", build_dijkstra_report())


if __name__ == "__main__":
    unittest.main()
